src/pyqasm/openpulse_visitor.py

```diff
@@ -0,0 +1,376 @@
+# Copyright 2025 qBraid
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=all
+# mypy: disable-error-code=union-attr
+"""
+Module defining OpenPulse Visitor.
+
+"""
+import logging
+from typing import Any
+
+import openpulse.ast as o_pulse_ast
+
+from pyqasm.elements import Frame
+from pyqasm.exceptions import (
+    raise_qasm3_error,
+)
+
+logger = logging.getLogger(__name__)
+logger.propagate = False
+
+
+class OpenPulseVisitor:
+    """A visitor for basic OpenPulse program elements.
+
+    This class is designed to traverse and interact with elements in an OpenPulse program.
+
+    Args:
+
+    """
+
+    def __init__(
+        self,
+        module,
+        check_only: bool = False,
+        is_def_cal: bool = False,
+    ):
+        self._module = module
+        self._check_only: bool = check_only
+        self._frames = list[Frame]
+        self._ports = list[o_pulse_ast.PortType]
+        self._waveforms = list[o_pulse_ast.WaveformType]
+        self._is_def_cal = is_def_cal
```

Comment:
We will most likely need some kind of reference to the scope of the `QasmVisitor` object to do the semantic validations. And `self._module = module` can then be derived from the visitor object